### PR TITLE
Fix Dynamo handling for random.random

### LINKED ISSUE #145446: [dynamo] fix graph break random.random
We used to not graph break on `random.random`, but we do now.

```python
import random

import torch


@torch.compile(backend="eager", fullgraph=True)
def fn(x):
    return x + random.random()


fn(torch.ones(5, 5))
```

This does not happen to other supported random functions - `randint`, `randrange`, `uniform`, listed in https://github.com/pytorch/pytorch/blob/d95a6babcc581ff06d1b914ee9f92c81b2e850e2/torch/_dynamo/variables/user_defined.py#L743.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185961

Dynamo already has special handling for supported stdlib random helpers: calls are recorded on OutputGraph.random_calls and their runtime values are fed back into the graph through RandomValueSource. random.randint, random.randrange, and random.uniform reached that path because they are Python bound methods and were wrapped as UserDefinedObjectVariable.

random.random is different: it is a bound C builtin on the module-level Random instance. VariableBuilder treated it as a generic callable, so trace_rules.lookup classified it as skipped before the existing random helper handling could run. With fullgraph=True this surfaced as a graph break for random.random even though the equivalent random helpers were supported.

Route supported stdlib random helper functions to UserDefinedObjectVariable before the generic callable trace-rule path, and share the supported-random predicate with the existing call_function handling. This keeps all supported helpers on the existing call_random_fn path instead of adding a one-off random.random workaround.

Fixes #145446
Generated by my agent

Test Plan:
python test/dynamo/test_unspec.py -k test_random_random_fullgraph
python test/dynamo/test_unspec.py -k random
git diff --check
git diff --cached --check
lintrunner -a (fails on unrelated existing torch/nn/modules/loss.py:1579 Pyrefly unexpected keyword argument `options`)
lintrunner torch/_dynamo/variables/builder.py torch/_dynamo/variables/user_defined.py test/dynamo/test_unspec.py (same unrelated Pyrefly error)

Benchmark Results:
This change adds a supported-random predicate check during Dynamo variable construction and has no compiled graph runtime effect. Microbenchmarking 1,000,000 unsupported-callable checks showed baseline false-return timings of [0.026731, 0.026803, 0.026769, 0.026823, 0.026875] seconds and new predicate timings of [0.083945, 0.085281, 0.083589, 0.083639, 0.083401] seconds, about 57 ns extra per check.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I reproduced the Dynamo-wrapped CI failures locally. They came from the supported-random predicate probing arbitrary objects with throwing __hash__ methods while Dynamo realized locals after graph breaks. I narrowed that predicate to the callable runtime types used by stdlib random helpers before set membership, and the previously failing representative tests now pass under PYTORCH_TEST_WITH_DYNAMO=1.

**Comment by jansel** [comment]
My agent says the TorchTitan failure is the known torchcomms process-group resolution issue tracked in pytorch/pytorch#186031 and pytorch/torchtitan#3493, not caused by this Dynamo random-handling change. I removed the extra ciflow/torchtitan label and refreshed the ghstack PR; the relevant CI is rerunning.

**Comment by jansel** [comment]
My agent says I addressed the Dynamo CPython CI failure by removing the stale expected-failure marker for CPython313 test_deque TestBasic.test_reverse; the PR fix makes that test pass because it uses random.random(). I also kept the unrelated TorchTitan torchcomms failure out of scope and am removing the ciflow/torchtitan label again so fresh CI focuses on the relevant lanes.

**Comment by jansel** [comment]
My agent says the reported DebertaV2ForMaskedLM Inductor CI failure was an eager Transformers import/JIT failure (eager_fail_to_run before Dynamo/Inductor execution) on the old base, not caused by this random.random handling change. I rebased and resubmitted on current main; the targeted random.random regression test and lintrunner -a pass.

**Comment by jansel** [comment]
My agent says the current TorchTitan failures did not reach the test body; both failed during dependency installation with a transient PyPI connection reset while fetching spmd_types metadata for torchao/torchcomms. I attempted to rerun the failed jobs, but the GitHub token available in this environment is returning HTTP 401 for write operations, so no code change is needed from this PR.

**Comment by jansel** [comment]
My agent says this is fixed by pytorch/pytorch#188235, which landed on main as 79de04232fa. I verified the original random.random fullgraph reproducer passes on current main.
